splink/comparison_helpers.py

```diff
@@ -0,0 +1,425 @@
+
+        ch.comparator_score_df(list, "string1", "string2")
+        ```
+    """
+    duckdb.connect()
+
+    pd.DataFrame(list)
+
+    sql = f"""
+        SELECT
+        {col1}, {col2},
+        levenshtein({col1}, {col2}) as levenshtein_distance,
+        damerau_levenshtein({col1}, {col2}) as damerau_levenshtein_distance,
+        ROUND(jaro_similarity({col1}, {col2}), {decimal_places}) as jaro_similarity,
+        ROUND(jaro_winkler_similarity(
+            {col1}, {col2}), {decimal_places}
+            ) as jaro_winkler_similarity,
+        ROUND(jaccard({col1}, {col2}), {decimal_places}) as jaccard_similarity
+        FROM df
+    """
+
+    return duckdb.sql(sql).df()
+
+
+def comparator_score_chart(
+    list, col1, col2, similarity_threshold=None, distance_threshold=None
+):
+    """Helper function returning a heatmap showing the sting similarity
+    scores and string distances for a list of strings.
+
+    Examples:
+        ```py
+        import splink.comparison_helpers as ch
+
+        list = {
+                "string1": ["Stephen", "Stephen","Stephen"],
+                "string2": ["Stephen", "Steven", "Stephan"],
+                }
+
+        ch.comparator_score_chart(list, "string1", "string2")
+        ```
+    """
+
+    df = comparator_score_df(list, col1, col2)
+
+    df["strings_to_compare"] = df["string1"] + ", " + df["string2"]
+
+    df_long = pd.melt(
+        df,
+        id_vars=["strings_to_compare"],
+        value_vars=[
+            "jaro_similarity",
+            "jaro_winkler_similarity",
+            "jaccard_similarity",
+            "levenshtein_distance",
+            "damerau_levenshtein_distance",
+        ],
+        var_name="comparator",
+        value_name="score",
+    )
+
+    similarity_title = "Heatmap of Similarity Scores"
+    distance_title = "Heatmap of Distance Scores"
+    graph_colour = "score"
+
+    if similarity_threshold and distance_threshold:
+        df_long["threshold_match"] = df_long.apply(
+            lambda row: threshold_match(
+                row["comparator"],
+                row["score"],
+                distance_threshold,
+                similarity_threshold,
+            ),
+            axis=1,
+        )
+        similarity_title = f"{similarity_title} with threshold {similarity_threshold}"
+        distance_title = f"{distance_title} with threshold {distance_threshold}"
+        graph_colour = "threshold_match"
+
+    # create similarity heatmap
+    heatmap_similarity = (
+        alt.Chart(
+            df_long[df_long["comparator"].str.contains("similarity")],
+            width=300,
+            height=300,
+        )
+        .mark_rect()
+        .encode(
+            x="comparator:O",
+            y="strings_to_compare:O",
+            color=alt.Color(
+                f"{graph_colour}:Q",
+                scale=alt.Scale(range=["red", "green"], domain=[0, 1]),
+            ),
+        )
+        .properties(title=similarity_title)
+    )
```

Comment:
Can we use `vegalite` for this?

You can certainly do [heatmaps](https://vega.github.io/vega-lite/examples/layer_text_heatmap.html) in there, I just don't know how to stitch multiple maps together.

Reply:
If we can, could you add the chart def to [splink/files/chart_defs](https://github.com/moj-analytical-services/splink/tree/master/splink/files/chart_defs).

I'd also suggest having this function output your final values in a json dict format and then adding this function to the charts section of our codebase (I can point that out if you're unclear on what I mean).

Reply:
Yea I can just save out the [chart as json](https://altair-viz.github.io/user_guide/saving_charts.html) and use that as a starting point

Reply:
For context, the reason behind using vega lite specs instead of Altair is to reduce dependencies. 

Specifically it means Splink works even on systems where Altair isn't (or can't be) installed.  

Only mentioning it because on the face of it it looks like an odd decision (since the Altair code is generally more succinct and readable)